contracts/AelinPool.sol

```diff
@@ -385,6 +443,103 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
+     * 2. certain amount of purchase tokens per wallet regardless of the number of qualified NFTs held
+     * 3. certain amount of Investment tokens per qualified NFT held
+     */
+
+    function purchasePoolTokensWithNft(NftArgs[] calldata nftArgs, uint256 _purchaseTokenAmount) external lock {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+
+        uint256 maxPurchaseTokenAmount;
+
+        for (uint256 i = 0; i < nftArgs.length; i++) {
+            NftArgs memory tmpNftArg = nftArgs[i];
+            address _collectionAddress = tmpNftArg.collectionAddress;
+            uint256[] memory _tokenIds = tmpNftArg.tokenIds;
+
+            NftData memory tmpNftData = nftCollectionDetails[_collectionAddress];
+            require(tmpNftData.collectionAddress == _collectionAddress, "collection not in the pool");
+
+            if (tmpNftData.purchaseAmountPerToken) {
+                maxPurchaseTokenAmount += tmpNftData.purchaseAmount * _tokenIds.length;
+            }
+
+            if (!tmpNftData.purchaseAmountPerToken && tmpNftData.purchaseAmount > 0) {
+                require(!nftWalletUsedForPurchase[_collectionAddress][msg.sender], "wallet already used for collection");
+                nftWalletUsedForPurchase[_collectionAddress][msg.sender] = true;
+                maxPurchaseTokenAmount += tmpNftData.purchaseAmount;
+            }
+
+            if (tmpNftData.purchaseAmount == 0) {
+                maxPurchaseTokenAmount = _purchaseTokenAmount;
+            }
+
+            if (NftCheck.supports721(_collectionAddress)) {
+                _purchasePoolWith721(_collectionAddress, _tokenIds);
+            }
+            if (NftCheck.supports1155(_collectionAddress)) {
+                _purchasePoolWith1155(_collectionAddress, _tokenIds, tmpNftData);
+            }
+            if (_collectionAddress == CRYPTO_PUNKS) {
+                _purchasePoolWithPunks(_collectionAddress, _tokenIds);
+            }
+        }
+
+        require(_purchaseTokenAmount <= maxPurchaseTokenAmount, "purchase amount should be less the max allocation");
+
+        uint256 amountBefore = IERC20(purchaseToken).balanceOf(address(this));
+        IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount);
+        uint256 amountAfter = IERC20(purchaseToken).balanceOf(address(this));
+        uint256 purchaseTokenAmount = amountAfter - amountBefore;
+
+        if (purchaseTokenCap > 0) {
+            uint256 totalPoolAfter = totalSupply() + purchaseTokenAmount;
+            require(totalPoolAfter <= purchaseTokenCap, "cap has been exceeded");
+            if (totalPoolAfter == purchaseTokenCap) {
+                purchaseExpiry = block.timestamp;
+            }
+        }
+
+        _mint(msg.sender, purchaseTokenAmount);
+    }
+
+    function _purchasePoolWith721(address _collectionAddress, uint256[] memory _tokenIds) internal {
```

Comment:
I don't like the name of this function. It implies that you are doing purchasing in the method but really you are just blacklisting NFTs

Reply:
will change it.